File: terrariabonker/content.py
"""Every item's default stats, read from the game's own template objects.

Stats like damage, defense and rarity are not in `Terraria.exe` in any readable form —
they are assigned by `Item.SetDefaults` at runtime. The authority is `ContentSamples`,
which holds one fully-populated `Item` per type.

`ContentSamples.ItemsByType` is a `Dictionary<int, Item>`, and walking a mono dictionary
means depending on the *runtime's* field layout rather than Terraria's — a failure mode the
build ledger would not catch cleanly. So this finds the templates a different way: scan the
writable regions for objects carrying the `Item` vtable (the scan `_template_block` already
does for one type), and pick out the template table by its shape.

The table gives itself away by being **one object per type**. Live items — inventories,
chests, dropped items — repeat types heavily, so a run of addresses holding roughly as many
objects as distinct types is a template table. Measured on a live game: 169,637 Item-shaped
objects in 463 ms, of which the one-to-one runs cover 6,162 distinct types, i.e. everything.
"""

from __future__ import annotations

from collections import Counter

import numpy as np

from terrariabonker.inventory import (ITEM_ACCESSORY, ITEM_AUTOREUSE, ITEM_BODY_SLOT,
                                      ITEM_BUFF_TYPE, ITEM_DAMAGE,
                                      ITEM_DEFENSE, ITEM_HEAD_SLOT, ITEM_HEAL_LIFE,
                                      ITEM_HEAL_MANA, ITEM_LEG_SLOT, ITEM_MAGIC, ITEM_MELEE,
                                      ITEM_PICK, ITEM_PREFIX, ITEM_RANGED, ITEM_RARE,
                                      ITEM_SUMMON,
                                      ITEM_TILEBOOST, ITEM_TYPE, ITEM_USE_ANIM,
                                      ITEM_USE_TIME)
from terrariabonker.npcs import (MAX_NPC_TYPE, NPC_BOSS, NPC_COLOR, NPC_DAMAGE,
                                 NPC_DEFENSE, NPC_HEIGHT, NPC_LIFE_MAX, NPC_NET_ID,
                                 NPC_OBJECT_SIZE, NPC_TOWN, NPC_TYPE, NPC_WIDTH)
from terrariabonker.recipes import ITEM_CREATE_TILE

# Addresses further apart than this start a new run. Chosen from live data: it merges the
# template table's chunks without swallowing the surrounding live-item heap.
CLUSTER_GAP = 0x400000
MAX_TYPE = 7000                 # sanity bound on a plausible ItemID


def _read_stats(buf: bytes, off: int) -> dict:
    def i32(field):
        return int.from_bytes(buf[off + field: off + field + 4], "little", signed=True)

    return {
        "type": i32(ITEM_TYPE),
        "damage": i32(ITEM_DAMAGE),
        "defense": i32(ITEM_DEFENSE),
        "rare": i32(ITEM_RARE),
        "pick": i32(ITEM_PICK),
        "use_time": i32(ITEM_USE_TIME),
        # Needed to tell a real edit from an item's own defaults (spec 038).
        "use_anim": i32(ITEM_USE_ANIM),
        "tile_boost": i32(ITEM_TILEBOOST),
        "auto_reuse": int(buf[off + ITEM_AUTOREUSE]),
        "create_tile": i32(ITEM_CREATE_TILE),
        "buff_type": i32(ITEM_BUFF_TYPE),
        "heal_life": i32(ITEM_HEAL_LIFE),
        "heal_mana": i32(ITEM_HEAL_MANA),
        "head_slot": i32(ITEM_HEAD_SLOT),
        "body_slot": i32(ITEM_BODY_SLOT),
        "leg_slot": i32(ITEM_LEG_SLOT),
        # Kept only to tell a modified copy from a pristine one; stripped before use.
        "prefix": int(buf[off + ITEM_PREFIX]),
        "accessory": bool(buf[off + ITEM_ACCESSORY]),
        "melee": bool(buf[off + ITEM_MELEE]),
        "ranged": bool(buf[off + ITEM_RANGED]),
        "magic": bool(buf[off + ITEM_MAGIC]),
        "summon": bool(buf[off + ITEM_SUMMON]),
    }


def _scan_objects(mem, vtable: int, span: int, read, accept) -> list[tuple[int, dict]]:
    """Every object carrying ``vtable``, as ``(address, stats)``, address-ordered.

    ``read(buf, off)`` turns one object into a stats dict; ``accept(stats)`` rejects the
    false positives a bare vtable match always produces (a stale pointer in the middle of
    some other structure reads as an object with an absurd type).
    """
    found: list[tuple[int, dict]] = []
    for start, end in mem.regions():
        buf = mem.read(start, end - start)
        n = len(buf) // 4
        if n < 4:
            continue
        arr = np.frombuffer(buf[: n * 4], dtype=np.uint32)
        for idx in np.where(arr == vtable)[0].tolist():
            off = idx * 4
            if off + span > len(buf):
                continue
            stats = read(buf, off)
            if accept(stats):
                found.append((start + off, stats))
    found.sort(key=lambda p: p[0])
    return found


def _template_table(found: list[tuple[int, dict]], key: str) -> dict[int, dict]:
    """Pick the template table out of every object of a class, keyed by ``key``.

    The table gives itself away by being one object per key: live objects — inventories,
    chests, the NPC slots — repeat their keys heavily, templates never do. Addresses are
    clustered into runs, and from each run we keep the keys that appear in it **exactly
    once**. Bigger runs are consulted first, so where a key appears in more than one run it
    is taken from the real table rather than from a chest that happens to look one-to-one.

    Judging keys rather than whole runs matters. Scoring a run as a whole and discarding it
    if it was not near enough to one-to-one lost real templates to whatever happened to be
    allocated beside them: the seven Moss Hornet variants sit next to seven default-state
    NPC objects, which made their run 14 objects for 8 distinct netIDs and threw away all
    seven. Per-key, the duplicated key drops out and its neighbours survive.
    """
    if not found:
        return {}
    runs: list[list[tuple[int, dict]]] = [[found[0]]]
    for addr, stats in found[1:]:
        if addr - runs[-1][-1][0] > CLUSTER_GAP:
            runs.append([])
        runs[-1].append((addr, stats))

    runs.sort(key=len, reverse=True)
    out: dict[int, dict] = {}
    for run in runs:
        seen = Counter(s[key] for _a, s in run)
        for _addr, stats in run:
            if seen[stats[key]] == 1:
                out.setdefault(stats[key], stats)
    return out


def _consensus(candidates: list[dict]) -> dict | None:
    """The stats a *pristine* copy of an item has (spec 039).

    Two rules, in order. A modifier changes damage and use time, so a prefixed copy is not
    a template: where any copy is unprefixed, only those are considered. Then take the
    most common stat tuple — a pristine template agrees with every other pristine copy of
    its type, while an edited one stands alone.

    Choosing by position instead let a live, modified item be returned as the template. It
    reported the maintainer's own edits as Boomstick's base stats, and spec 038 then used
    those as the baseline for deciding which saved edits were redundant, destroying eight
    of them.
    """
    clean = [c for c in candidates if not c.get("prefix")] or candidates
    if len(clean) == 1:
        return clean[0]
    keys = tuple(k for k in clean[0] if k != "prefix")
    tally = Counter(tuple(c[k] for k in keys) for c in clean)
    winner = tally.most_common(1)[0][0]
    for c in clean:
        if tuple(c[k] for k in keys) == winner:
            return c
    return clean[0]


def find_item_templates(mem, vtable: int, exclude=()) -> dict[int, dict]:
    """``{type: stats}`` for every item the game has a template for.

    ``exclude`` is the addresses of objects known to be the player's own — the ones this
    program edits, and so the likeliest to be mistaken for a template.
    """
    span = max(ITEM_TYPE, ITEM_DAMAGE, ITEM_DEFENSE, ITEM_RARE, ITEM_PICK, ITEM_USE_TIME,
               ITEM_CREATE_TILE, ITEM_ACCESSORY, ITEM_SUMMON, ITEM_LEG_SLOT,
               ITEM_HEAL_MANA, ITEM_BUFF_TYPE, ITEM_USE_ANIM, ITEM_TILEBOOST,
               ITEM_AUTOREUSE, ITEM_PREFIX) + 4
    skip = set(exclude or ())
    found = _scan_objects(mem, vtable, span, _read_stats,
                          lambda s: 0 < s["type"] < MAX_TYPE)
    by_type: dict[int, list[dict]] = {}
    for addr, stats in found:
        if addr in skip:
            continue
        by_type.setdefault(stats["type"], []).append(stats)
    out = {}
    for t, cands in by_type.items():
        got = _consensus(cands)
        if got is not None:
            out[t] = {k: v for k, v in got.items() if k != "prefix"}
    return out


def _read_npc_stats(buf: bytes, off: int) -> dict:
    def i32(field):
        return int.from_bytes(buf[off + field: off + field + 4], "little", signed=True)

    return {
        "type": i32(NPC_TYPE),
        "net_id": i32(NPC_NET_ID),
        "life": i32(NPC_LIFE_MAX),
        "damage": i32(NPC_DAMAGE),
        "defense": i32(NPC_DEFENSE),
        "width": i32(NPC_WIDTH),
        "height": i32(NPC_HEIGHT),
        "boss": bool(buf[off + NPC_BOSS]),
        "town": bool(buf[off + NPC_TOWN]),
        # The tint the game paints a neutral sheet with; (0,0,0,0) means "no tint".
        "color": list(buf[off + NPC_COLOR: off + NPC_COLOR + 4]),
    }


def find_npc_templates(mem, vtable: int, exclude=()) -> dict[int, dict]:
    """``{net_id: stats}`` for every NPC the game has a template for.

    Keyed on ``netID`` rather than ``type`` because ContentSamples is: the variant entries
    (the coloured slimes and so on) share a type and are told apart only by a negative
    netID, and those are exactly the names ``data/npcs.json`` carries.

    Reading the templates rather than ``Main.npc[]`` is not a detail — a live NPC's stats
    have been scaled by the world's difficulty, so a Blue Slime in an expert world reads
    60 life where its template says 25.
    """
    skip = set(exclude or ())
    found = _scan_objects(mem, vtable, NPC_OBJECT_SIZE, _read_npc_stats,
                          lambda s: -MAX_NPC_TYPE < s["net_id"] < MAX_NPC_TYPE
                          and 0 <= s["type"] < MAX_NPC_TYPE)
    # The NPCs in the world have had their stats scaled by the world's difficulty, so they
    # are not templates however they happen to be laid out (spec 039).
    found = [(a, st) for a, st in found if a not in skip]
    return _template_table(found, "net_id")


def item_kind(stats: dict) -> str:
    """A one-word category for the compendium, from the fields the project already reads.

    Ordered most-specific first: an accessory that also has damage is still an accessory,
    and a pickaxe that deals damage is still a tool.
    """
    if stats.get("accessory"):
        return "Accessory"
    if max(stats.get("head_slot", -1), stats.get("body_slot", -1),
           stats.get("leg_slot", -1)) >= 0:
        return "Armor"          # includes vanity: a slot with no defense is still worn
    if stats.get("pick", 0) > 0:
        return "Tool"
    if stats.get("damage", 0) > 0:
        # Before the buff check: a summon staff grants its minion as a buff, and is a
        # weapon, not a potion.
        if stats.get("summon"):
            return "Summon"
        if stats.get("magic"):
            return "Magic"
        if stats.get("ranged"):
            return "Ranged"
        return "Weapon"
    if (stats.get("heal_life", 0) > 0 or stats.get("heal_mana", 0) > 0
            or stats.get("buff_type", 0) > 0):
        # Buff potions carry no healLife/healMana at all, which is why filtering on those
        # alone showed only restoratives. Food grants a buff too and lands here.
        return "Potion"
    if stats.get("defense", 0) > 0:
        return "Armor"
    if stats.get("create_tile", -1) >= 0:
        return "Block"
    return "Material"


def npc_kind(stats: dict) -> str:
    """Boss / town NPC / critter / monster, from the NPC template's own fields.

    ``boss`` and ``townNPC`` are real flags on the object. "Critter" is not — nothing on
    the template says "critter" in a way this project could confirm, so it is defined by
    what is observable: an NPC that deals no damage. That puts the bunnies and birds
    (5 life, 0 damage) where a player expects them, and misfiles the Target Dummy, which
    is an acceptable price for not inventing a flag.
    """
    if stats.get("boss"):
        return "Boss"
    if stats.get("town"):
        return "Town NPC"
    if stats.get("damage", 0) <= 0:
        return "Critter"
    return "Monster"


def wiki_url(name: str) -> str:
    """The official wiki article for a display name.

    Opened in the user's browser rather than fetched: the app makes no network requests.
    Article titles use underscores; a name that redirects or misses lands on the wiki's own
    search, which is an acceptable outcome for the handful that differ.
    """
    slug = "_".join(name.split())
    return "https://terraria.wiki.gg/wiki/" + slug

File: internal/content/content.go
/*
Package content is every item's and NPC's default stats, read from the game's own
template objects.

Stats like damage, defense and rarity are not in the game's executable in any
readable form -- they are assigned at runtime. The authority is the game's sample
collection, which holds one fully-populated object per type.

That collection is a managed dictionary, and walking one means depending on the
*runtime's* field layout rather than the game's, which is a failure the build
ledger would not catch cleanly. So the templates are found a different way: scan
the writable memory for objects carrying the right vtable, and pick the template
table out by its shape.

The table gives itself away by being **one object per type**. Live objects --
inventories, chests, dropped items -- repeat types heavily, so a run of addresses
holding roughly as many objects as distinct types is a template table. Measured
on a live game: 169,637 item-shaped objects, of which the one-to-one runs cover
6,162 distinct types, which is all of them.

Ported from terrariabonker/content.py (spec 051, step 5).
*/
package content

import (
	"encoding/binary"
	"sort"
	"strings"

	"github.com/ushineko/terrariabonker/internal/layout"
	"github.com/ushineko/terrariabonker/internal/proc"
)

/*
ClusterGap is how far apart two addresses have to be to start a new run.

Chosen from live data: it merges the template table's chunks without swallowing
the surrounding heap of live items.
*/
const ClusterGap = 0x400000

// MaxType bounds a plausible item id: a stale pointer read as an object gives an
// absurd one.
const MaxType = 7000

// Mem is the memory the templates are scanned out of.
type Mem interface {
	Regions() []proc.Region
	Read(addr uint32, size int) []byte
}

// ItemStats is one item's defaults.
type ItemStats struct {
	Type    int32 `json:"type"`
	Damage  int32 `json:"damage"`
	Defense int32 `json:"defense"`
	Rare    int32 `json:"rare"`
	Pick    int32 `json:"pick"`
	UseTime int32 `json:"use_time"`
	// UseAnim is needed to tell a real edit from an item's own defaults.
	UseAnim    int32 `json:"use_anim"`
	TileBoost  int32 `json:"tile_boost"`
	AutoReuse  int32 `json:"auto_reuse"`
	CreateTile int32 `json:"create_tile"`
	BuffType   int32 `json:"buff_type"`
	HealLife   int32 `json:"heal_life"`
	HealMana   int32 `json:"heal_mana"`
	HeadSlot   int32 `json:"head_slot"`
	BodySlot   int32 `json:"body_slot"`
	LegSlot    int32 `json:"leg_slot"`
	/*
		Prefix is kept only to tell a modified copy from a pristine one, and
		never leaves this package: a template by definition carries none, so
		reporting it would be reporting a field that is always zero and inviting
		somebody to restore it.
	*/
	Prefix    int32 `json:"-"`
	Accessory bool  `json:"accessory"`
	Melee     bool  `json:"melee"`
	Ranged    bool  `json:"ranged"`
	Magic     bool  `json:"magic"`
	Summon    bool  `json:"summon"`
}

// found is one object the scan turned up.
type found[T any] struct {
	addr  uint32
	stats T
}

/*
scanObjects is every object carrying a vtable, in address order.

read turns one object into its stats; accept rejects the false positives a bare
vtable match always produces, because a stale pointer in the middle of some other
structure reads as an object with an absurd type.
*/
func scanObjects[T any](mem Mem, vtable uint32, span int,
	read func([]byte, int) T, accept func(T) bool) []found[T] {
	var out []found[T]
	for _, region := range mem.Regions() {
		buf := mem.Read(region.Start, region.Size())
		for off := 0; off+4 <= len(buf); off += 4 {
			if binary.LittleEndian.Uint32(buf[off:]) != vtable {
				continue
			}
			if off+span > len(buf) {
				continue
			}
			stats := read(buf, off)
			if accept(stats) {
				out = append(out, found[T]{addr: region.Start + uint32(off), stats: stats}) //nolint:gosec // an offset in a 32-bit region
			}
		}
	}
	sort.Slice(out, func(i, j int) bool { return out[i].addr < out[j].addr })
	return out
}

/*
templateTable picks the template table out of every object of a class, keyed by
whatever tells one template from another.

The table gives itself away by being one object per key: live objects repeat
their keys heavily, templates never do. Addresses are clustered into runs, and
from each run the keys appearing in it **exactly once** are kept. Bigger runs are
consulted first, so a key appearing in more than one run is taken from the real
table rather than from a chest that happens to look one-to-one.

Judging keys rather than whole runs matters. Scoring a run as a whole and
discarding it when it was not near enough to one-to-one lost real templates to
whatever happened to be allocated beside them: seven variants of one NPC sit next
to seven default-state objects, which made their run fourteen objects for eight
distinct keys and threw away all seven. Per key, the duplicated key drops out and
its neighbours survive.
*/
func templateTable[T any](objects []found[T], key func(T) int32) map[int32]T {
	out := map[int32]T{}
	if len(objects) == 0 {
		return out
	}
	runs := [][]found[T]{{objects[0]}}
	for _, o := range objects[1:] {
		last := runs[len(runs)-1]
		if o.addr-last[len(last)-1].addr > ClusterGap {
			runs = append(runs, nil)
		}
		runs[len(runs)-1] = append(runs[len(runs)-1], o)
	}
	sort.SliceStable(runs, func(i, j int) bool { return len(runs[i]) > len(runs[j]) })

	for _, run := range runs {
		seen := map[int32]int{}
		for _, o := range run {
			seen[key(o.stats)]++
		}
		for _, o := range run {
			k := key(o.stats)
			if seen[k] != 1 {
				continue
			}
			if _, already := out[k]; !already {
				out[k] = o.stats
			}
		}
	}
	return out
}

/*
consensus is the stats a *pristine* copy of an item has.

Two rules, in order. A modifier changes damage and use time, so a prefixed copy
is not a template: where any copy is unprefixed, only those are considered. Then
the most common stat tuple wins -- a pristine template agrees with every other
pristine copy of its type, while an edited one stands alone.

Choosing by position instead let a live, modified item be returned as the
template. It reported the maintainer's own edits as one weapon's base stats, and
those were then used as the baseline for deciding which saved edits were
redundant, destroying eight of them.
*/
func consensus(candidates []ItemStats) (ItemStats, bool) {
	if len(candidates) == 0 {
		return ItemStats{}, false
	}
	var clean []ItemStats
	for _, c := range candidates {
		if c.Prefix == 0 {
			clean = append(clean, c)
		}
	}
	if len(clean) == 0 {
		clean = candidates
	}
	if len(clean) == 1 {
		return clean[0], true
	}
	tally := map[ItemStats]int{}
	for _, c := range clean {
		tally[withoutPrefix(c)]++
	}
	/*
		Walked in the order the copies were found rather than over the tally, so
		that equally-common candidates are broken the same way every time and the
		same way the Python breaks them -- first seen wins.

		Ranging over the tally instead reads a Go map, whose order is deliberately
		random: two copies of one type that disagree would then produce a
		different template on different runs of the same program, against the
		same game. That is worse than picking the wrong one, because it cannot be
		reproduced.
	*/
	best, most := clean[0], 0
	for _, c := range clean {
		if n := tally[withoutPrefix(c)]; n > most {
			best, most = c, n
		}
	}
	return best, true
}

// withoutPrefix is a copy with the modifier tier cleared, which is the only
// field a template is allowed to disagree on.
func withoutPrefix(s ItemStats) ItemStats {
	s.Prefix = 0
	return s
}

// itemSpan is how much of an item object has to be readable before its fields
// can be trusted.
var itemSpan = furthest(layout.ItemType, layout.ItemDamage, layout.ItemDefense,
	layout.ItemRare, layout.ItemPick, layout.ItemUseTime, layout.ItemCreateTile,
	layout.ItemAccessory, layout.ItemSummon, layout.ItemLegSlot, layout.ItemHealMana,
	layout.ItemBuffType, layout.ItemUseAnim, layout.ItemTileBoost, layout.ItemAutoReuse,
	layout.ItemPrefix) + 4

// furthest is the largest of some offsets.
func furthest(offs ...int) int {
	most := 0
	for _, o := range offs {
		if o > most {
			most = o
		}
	}
	return most
}

/*
FindItemTemplates is every item the game has a template for.

exclude is the addresses of objects known to be the player's own -- the ones this
program edits, and so the likeliest to be mistaken for a template.
*/
func FindItemTemplates(mem Mem, vtable uint32, exclude map[uint32]bool) map[int32]ItemStats {
	objects := scanObjects(mem, vtable, itemSpan, readItemStats,
		func(s ItemStats) bool { return s.Type > 0 && s.Type < MaxType })

	byType := map[int32][]ItemStats{}
	for _, o := range objects {
		if exclude[o.addr] {
			continue
		}
		byType[o.stats.Type] = append(byType[o.stats.Type], o.stats)
	}
	out := make(map[int32]ItemStats, len(byType))
	for t, candidates := range byType {
		if got, ok := consensus(candidates); ok {
			out[t] = withoutPrefix(got)
		}
	}
	return out
}

// readItemStats is one item object's fields.
func readItemStats(buf []byte, off int) ItemStats {
	i32 := func(field int) int32 {
		return int32(binary.LittleEndian.Uint32(buf[off+field:])) //nolint:gosec // a field, as its bits
	}
	return ItemStats{
		Type: i32(layout.ItemType), Damage: i32(layout.ItemDamage),
		Defense: i32(layout.ItemDefense), Rare: i32(layout.ItemRare),
		Pick: i32(layout.ItemPick), UseTime: i32(layout.ItemUseTime),
		UseAnim: i32(layout.ItemUseAnim), TileBoost: i32(layout.ItemTileBoost),
		AutoReuse:  int32(buf[off+layout.ItemAutoReuse]),
		CreateTile: i32(layout.ItemCreateTile), BuffType: i32(layout.ItemBuffType),
		HealLife: i32(layout.ItemHealLife), HealMana: i32(layout.ItemHealMana),
		HeadSlot: i32(layout.ItemHeadSlot), BodySlot: i32(layout.ItemBodySlot),
		LegSlot: i32(layout.ItemLegSlot), Prefix: int32(buf[off+layout.ItemPrefix]),
		Accessory: buf[off+layout.ItemAccessory] != 0,
		Melee:     buf[off+layout.ItemMelee] != 0,
		Ranged:    buf[off+layout.ItemRanged] != 0,
		Magic:     buf[off+layout.ItemMagic] != 0,
		Summon:    buf[off+layout.ItemSummon] != 0,
	}
}

// NPCStats is one NPC's defaults.
type NPCStats struct {
	Type    int32 `json:"type"`
	NetID   int32 `json:"net_id"`
	Life    int32 `json:"life"`
	Damage  int32 `json:"damage"`
	Defense int32 `json:"defense"`
	Width   int32 `json:"width"`
	Height  int32 `json:"height"`
	Boss    bool  `json:"boss"`
	Town    bool  `json:"town"`
	// Color is the tint the game paints a neutral sheet with. All zero means no
	// tint.
	Color [4]byte `json:"color"`
}

// readNPCStats is one NPC object's fields.
func readNPCStats(buf []byte, off int) NPCStats {
	i32 := func(field int) int32 {
		return int32(binary.LittleEndian.Uint32(buf[off+field:])) //nolint:gosec // a field, as its bits
	}
	var color [4]byte
	copy(color[:], buf[off+layout.NPCColor:off+layout.NPCColor+4])
	return NPCStats{
		Type: i32(layout.NPCType), NetID: i32(layout.NPCNetID),
		Life: i32(layout.NPCLifeMax), Damage: i32(layout.NPCDamage),
		Defense: i32(layout.NPCDefense), Width: i32(layout.NPCWidth),
		Height: i32(layout.NPCHeight),
		Boss:   buf[off+layout.NPCBoss] != 0, Town: buf[off+layout.NPCTown] != 0,
		Color: color,
	}
}

/*
FindNPCTemplates is every NPC the game has a template for, keyed by net id.

Keyed on the net id rather than the type because the game's own collection is:
the variants share a type and are told apart only by a negative net id, and those
are exactly the names the bundled table carries.

Reading the templates rather than the NPCs in the world is not a detail -- a live
NPC's stats have been scaled by the world's difficulty, so a Blue Slime in an
expert world reads 60 life where its template says 25.
*/
func FindNPCTemplates(mem Mem, vtable uint32, exclude map[uint32]bool) map[int32]NPCStats {
	objects := scanObjects(mem, vtable, layout.NPCObjectSize, readNPCStats,
		func(s NPCStats) bool {
			return s.NetID > -layout.MaxNPCType && s.NetID < layout.MaxNPCType &&
				s.Type >= 0 && s.Type < layout.MaxNPCType
		})
	kept := objects[:0]
	for _, o := range objects {
		if !exclude[o.addr] {
			kept = append(kept, o)
		}
	}
	return templateTable(kept, func(s NPCStats) int32 { return s.NetID })
}

/*
ItemKind is a one-word category for the compendium, from the fields this project
already reads.

Ordered most-specific first: an accessory that also deals damage is still an
accessory, and a pickaxe that deals damage is still a tool.
*/
func ItemKind(s ItemStats) string {
	switch {
	case s.Accessory:
		return "Accessory"
	// A slot with no defense is still worn, so vanity counts as armour.
	case s.HeadSlot >= 0 || s.BodySlot >= 0 || s.LegSlot >= 0:
		return "Armor"
	case s.Pick > 0:
		return "Tool"
	case s.Damage > 0:
		// Before the buff check: a summon staff grants its minion as a buff and
		// is a weapon, not a potion.
		switch {
		case s.Summon:
			return "Summon"
		case s.Magic:
			return "Magic"
		case s.Ranged:
			return "Ranged"
		}
		return "Weapon"
	// Buff potions carry no healing at all, which is why filtering on healing
	// alone showed only restoratives. Food grants a buff too and lands here.
	case s.HealLife > 0 || s.HealMana > 0 || s.BuffType > 0:
		return "Potion"
	case s.Defense > 0:
		return "Armor"
	case s.CreateTile >= 0:
		return "Block"
	}
	return "Material"
}

/*
NPCKind is boss, town NPC, critter or monster, from the template's own fields.

Boss and town are real flags on the object. "Critter" is not -- nothing on the
template says critter in a way this project could confirm, so it is defined by
what is observable: an NPC that deals no damage. That puts the bunnies and birds
where a player expects them, and misfiles the target dummy, which is an
acceptable price for not inventing a flag.
*/
func NPCKind(s NPCStats) string {
	switch {
	case s.Boss:
		return "Boss"
	case s.Town:
		return "Town NPC"
	case s.Damage <= 0:
		return "Critter"
	}
	return "Monster"
}

/*
WikiURL is the official wiki article for a display name.

Opened in the user's browser rather than fetched: this program makes no network
requests. Article titles use underscores, and a name that redirects or misses
lands on the wiki's own search, which is an acceptable outcome for the handful
that differ.
*/
func WikiURL(name string) string {
	return "https://terraria.wiki.gg/wiki/" + strings.Join(strings.Fields(name), "_")
}
